src/endless/Analytics.cpp

```diff
@@ -114,43 +115,55 @@ CString AnalyticsRequestThread::GetUserAgent()
 	return ret;
 }
 
+// In debug builds, CHttpConnection and CHttpFile print a warning (in debug
+// builds) from the destructor if they've not been explicitly Close()d.
+template <class T>
+void CloseDelete(T *x)
+{
+	x->Close();
+	delete x;
+}
+template <class T>
+using close_ptr = std::unique_ptr<T, decltype(&CloseDelete<T>)>;
+
 void AnalyticsRequestThread::SendRequest(const CStringA &bodyUtf8)
 {
 	try {
 		CString headers = _T("Content-type: application/x-www-form-urlencoded");
-		CHttpConnection *conn = m_session.GetHttpConnection(_T(SERVER_NAME), (INTERNET_PORT)SERVER_PORT);
+		close_ptr<CHttpConnection> conn(
+			m_session.GetHttpConnection(_T(SERVER_NAME), (INTERNET_PORT)SERVER_PORT),
+			&CloseDelete<CHttpConnection>);
 		CString path = m_debug ? SERVER_DEBUG_PATH : SERVER_PATH;
 		const DWORD flags = INTERNET_FLAG_EXISTING_CONNECT
 			| INTERNET_FLAG_KEEP_CONNECTION
 			| INTERNET_FLAG_SECURE
 			| INTERNET_FLAG_IGNORE_REDIRECT_TO_HTTP
 			| INTERNET_FLAG_NO_COOKIES
 			| INTERNET_FLAG_NO_UI;
-		CHttpFile *file = conn->OpenRequest(CHttpConnection::HTTP_VERB_POST, path,
-			/* Referer */ NULL, /* context */ 1, /* Accept */ NULL,
-			_T("HTTP/1.1"), flags);
+		close_ptr<CHttpFile> file(
+			conn->OpenRequest(CHttpConnection::HTTP_VERB_POST, path,
+				/* Referer */ NULL, /* context */ 1, /* Accept */ NULL,
+				_T("HTTP/1.1"), flags),
+			&CloseDelete<CHttpFile>);
 		if (file) {
 			file->SendRequest(headers, (LPVOID)(LPCSTR)bodyUtf8, bodyUtf8.GetLength());
 			uprintf("Analytics req: %s\n", (LPCSTR)bodyUtf8);
 			DWORD responseCode = 0;
 			IFFALSE_PRINTERROR(file->QueryInfoStatusCode(responseCode), "QueryInfoStatusCode failed");
 			uprintf("Analytics: response code %d", responseCode);
 			if (m_debug) {
-				HandleDebugResponse(file);
+				HandleDebugResponse(file.get());
 			}
-			file->Close();
-			delete file;
 		}
 		else {
 			uprintf("Analytics req failed\n");
 		}
-		conn->Close();
-		delete conn;
 	}
 	catch (CInternetException *ex) {
 		TCHAR err[256];
 		ex->GetErrorMessage(err, 256);
 		uprintf("Analytics req error: %ls\n", (LPCTSTR)err);
+		ex->Delete();
```

Comment:
@ramcq https://msdn.microsoft.com/en-us/library/0e5twxsh.aspx

> Your exception handlers must delete exception objects they handle, because failure to delete the exception causes a memory leak whenever that code catches an exception.